src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -19,6 +21,153 @@ namespace Microsoft.CodeAnalysis.CSharp
+            };
+
+            public override string? ToString() => Parameter is { } p
+                ? p.ToString()
+                : Argument.ToString();
+        }
+
+        /// <summary>
+        /// Represents an argument being analyzed for escape analysis purposes. This represents the
+        /// argument as written. For example a `ref x` will only be represented by a single 
+        /// <see cref="EscapeArgument"/>
+        /// </summary>
+        internal readonly struct EscapeArgument
+        {
+            /// <summary>
+            /// This will be null in cases like arglist or a function pointer receiver.
+            /// </summary>
+            internal ParameterSymbol? Parameter { get; }
+
+            internal BoundExpression Argument { get; }
+
+            internal RefKind RefKind { get; }
+
+            internal EscapeArgument(ParameterSymbol? parameter, BoundExpression argument, RefKind refKind, bool isArgList = false)
+            {
+                Debug.Assert(!isArgList || parameter is null);
+                Argument = argument;
+                Parameter = parameter;
+                RefKind = refKind;
+            }
+
+            public void Deconstruct(out ParameterSymbol? parameter, out BoundExpression argument, out RefKind refKind)
+            {
+                parameter = Parameter;
+                argument = Argument;
+                refKind = RefKind;
+            }
+
+            public override string? ToString() => Parameter is { } p
+                ? p.ToString()
+                : Argument.ToString();
+        }
+
+        /// <summary>
+        /// Represents a value being analyzed for escape analysis purposes. This represents the value 
+        /// as it contributes to escape analysis which means arguments can show up multiple times. For
+        /// example `ref x` will be represented as both a val and ref escape
+        /// </summary>
+        internal readonly struct EscapeValue
+        {
+            /// <summary>
+            /// This will be null in cases like arglist or a function pointer receiver.
+            /// </summary>
+            internal ParameterSymbol? Parameter { get; }
+
+            internal BoundExpression Argument { get; }
+
+            /// <summary>
+            /// This differentiates whethere the value can only escape to <see cref="EscapeLevel.CallingMethod"/>
+            /// locations or to any return spot.
+            /// </summary>
+            internal EscapeLevel EscapeLevel { get; }
+
+            internal bool IsRefEscape { get; }
+
+            internal EscapeValue(ParameterSymbol parameter, BoundExpression argument, EscapeLevel escapeLevel, bool isRefEscape)
+            {
+                Argument = argument;
+                Parameter = parameter;
+                EscapeLevel = escapeLevel;
+                IsRefEscape = isRefEscape;
+            }
+
+            internal EscapeValue(BoundExpression argument, EscapeLevel escapeLevel, bool isRefEscape)
```

Comment:
>EscapeValue [](http://example.com/codeflow?start=21&length=11)

Consider removing this constructor and passing `parameter: null` explicitly at the call sites. Or if we really need multiple constructors, have this constructor call the constructor above.